klippa/src/serialize.rs

```diff
@@ -303,6 +454,86 @@ impl Serializer {
             .set_with_hash(to.0, hash, to.1, &self.data, &self.object_pool);
     }
 
+    fn resolve_links(&mut self) {
+        if self.in_error() {
+            return;
+        }
+
+        if self.current.is_some() || self.packed.is_empty() {
+            self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+            return;
+        }
+
+        let mut offset_links = Vec::new();
+        for parent_obj_idx in self.packed.iter() {
+            let Some(parent_obj) = self.object_pool.get_obj(*parent_obj_idx) else {
+                self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                return;
+            };
+
+            for link in parent_obj.real_links.iter() {
+                let Some(child_pool_idx) = self.packed.get(link.objidx) else {
+                    self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                    return;
+                };
+                let Some(child_obj) = self.object_pool.get_obj(*child_pool_idx) else {
+                    self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                    return;
+                };
+
+                let offset = match link.whence {
+                    OffsetWhence::Head => child_obj.head - parent_obj.head,
+                    OffsetWhence::Tail => child_obj.head - parent_obj.tail,
+                    OffsetWhence::Absolute => self.head - self.start + child_obj.head - self.tail,
+                };
+
+                if offset < link.bias as usize {
+                    self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                    return;
+                }
+
+                let offset = offset - link.bias as usize;
+                let offset_start_pos = parent_obj.head + link.position as usize;
+                offset_links.push((offset_start_pos, link.width, offset));
+            }
+        }
+
+        for (offset_start_pos, offset_width, offset) in offset_links.iter() {
+            self.assign_offset(*offset_start_pos, *offset_width, *offset);
+        }
+    }
+
+    //TODO: take care of signed offset
+    fn assign_offset(&mut self, offset_start_pos: usize, link_width: LinkWidth, offset: usize) {
+        let (offset_width, max_offset) = match link_width {
```

Comment:
maybe move this to a method on `LinkWidth`?

Reply:
it updates some values in the serializer.data so I'd like to keep it as an serializer method